stan/math/prim/err/check_matching_dims.hpp

```diff
@@ -26,20 +26,11 @@ namespace math {
 template <typename T1, typename T2, require_all_not_matrix_t<T1, T2>* = nullptr>
 inline void check_matching_dims(const char* function, const char* name1,
                                 const T1& y1, const char* name2, const T2& y2) {
+  STAN_NO_RANGE_AND_SIZE_CHECK
   std::vector<int> y1_d = dims(y1);
   std::vector<int> y2_d = dims(y2);
   bool error = false;
-  if (y1_d.size() != y2_d.size()) {
-    error = true;
-  } else {
-    for (int i = 0; i < y1_d.size(); i++) {
-      if (y1_d[i] != y2_d[i]) {
-        error = true;
-        break;
-      }
-    }
-  }
-  if (error) {
+  auto error_throw = [&]() STAN_COLD_PATH {
```

Comment:
In this function this should not be in lambda and cold path. This is the check itself, which is executed every time this is called. We only want construction of the error message in the cold path lambda.

Reply:
For this the lambda ends in `invalid_argument()` so if the lambda is called this will throw. Isn't the check itself here

```cpp
  std::vector<int> y1_d = dims(y1);
  std::vector<int> y2_d = dims(y2);
// yada yada...
  if (y1_d.size() != y2_d.size()) {
    error_throw();
  } else {
    for (int i = 0; i < y1_d.size(); i++) {
      if (y1_d[i] != y2_d[i]) {
        error_throw();
      }
    }
  }
```

Reply:
I guess I was not clear. I consider the check to be the branch between the error path and the ok path. Regardless we only want to put cold in the cold path if it is only executed if the check is failing. This code you put in the cold path here needs to be executed every time this `check_matching_dims` is called.